### DISCUSSION TITLE
Mismatch between the transformed image and the label

### DISCUSSION THREAD

**Discussion by wenwenzhang18** [TorchIO-project/torchio#793] [Q&A]
When elastic transformation or affine transformation is performed on the image and label at the same time, there is a mismatch between the transformed image and the label.

@fepegar ![image](https://user-images.githubusercontent.com/98071160/150266482-11f56cd6-c53d-412e-976b-1a4a182e3280.png)

**Comment by fepegar** [comment]
Can you please share the data and a minimal working example to reproduce this issue?

**Comment by wenwenzhang18** [comment]
[testData.zip](https://github.com/fepegar/torchio/files/7909959/testData.zip)
![image](https://user-images.githubusercontent.com/98071160/150457339-1273d302-5672-4460-9f24-9a4068299e72.png)
This is the data and code, at the same time I found that different images of the same subject will be transformed in the same way, but the label is not like this

**Reply by fepegar** [reply]
Can you please paste the code here using Python syntax highlighting, instead of uploading a screenshot? That is so I can copy the text and run the code.

**Comment by wenwenzhang18** [comment]
```python
import torch
import torchio as tio
from torch.utils.data import DataLoader   
import SimpleITK as itk
subject_a = tio.Subject(
    t1=tio.ScalarImage(r'E:\Data\Anatomy\mhd\T1\TrainData\T1_1.mhd'),
    label = tio.LabelMap(r'E:\Data\Anatomy\mhd\T1\TrainData\MaskT1_1.mhd'),
    label2 = tio.LabelMap(r'E:\Data\Anatomy\mhd\T1\TrainData\MaskT1_1.mhd')
)
subjects_list = [subject_a]
rescale = tio.RescaleIntensity(out_min_max=(0, 1))
spatial = tio.OneOf({
        tio.RandomAffine(
            scales=(0.5,2), 
            degrees=(10, 10, 10)):1,
        tio.RandomElasticDeformation(
            num_control_points=(20,20,20),
            max_displacement=10,
            locked_borders=2): 0,
        tio.RandomBlur():0,
        tio.RandomMotion():0,
        #tio.BiasField():1,
        #tio.RandomSwap(200,2):1
        tio.RandomLabelsToImage():0
    },
)
transforms = [rescale, spatial]
transform = tio.Compose(transforms)

# SubjectsDataset is a subclass of torch.data.utils.Dataset
subjects_dataset = tio.SubjectsDataset(subjects_list, transform=transform)

a_tranform = subjects_dataset[0]["t1"]["data"]
a_mask_tranform = subjects_dataset[0]["label"]["data"]


a_Tran = a_tranform.numpy()[0]
a_mask_Tran = a_mask_tranform.numpy()[0]


a_img = itk.GetImageFromArray(a_Tran.transpose(2,1,0))
a_mask_img = itk.GetImageFromArray(a_mask_Tran.transpose(2,1,0))

a_img.SetSpacing([0.80000001192092896,0.9375,0.9375])
a_mask_img.SetSpacing([0.80000001192092896,0.9375,0.9375])
itk.WriteImage(a_img,r"C:\a_img.mhd")
itk.WriteImage(a_mask_img,r"C:\a_mask_img.mhd")

##############

a_mask2_tranform = subjects_dataset[0]["label2"]["data"]
a_mask2_Tran = a_mask2_tranform.numpy()[0]
a_mask2_img = itk.GetImageFromArray(a_mask2_Tran.transpose(2,1,0))
a_mask2_img.SetSpacing([0.80000001192092896,0.9375,0.9375])
itk.WriteImage(a_mask2_img,r"C:\a_mask2_img.mhd")
```

**Reply by fepegar** [reply]
Thanks. When you do

```python
a_tranform = subjects_dataset[0]["t1"]["data"]
a_mask_tranform = subjects_dataset[0]["label"]["data"]
```

you call the `__getitem__` method of `subjects_dataset` twice, which applies two random transforms to the first subject in the dataset. You should do

```python
a_subject = subjects_dataset[0]
a_tranform = a_subject["t1"]["data"]
a_mask_tranform = a_subject["label"]["data"]
```

**Reply by fepegar** [reply]
By the way, note there is a method called `as_sitk` in the `Image` class, so you can do `itk.WriteImage(a_subject.t1.as_sitk(), r"C:\a_img.mhd")`.

**Reply by wenwenzhang18** [reply]
Thanks, the problem is solved, best wish to you.
